hotshot-state-prover/src/service.rs

```diff
@@ -315,7 +340,39 @@ pub async fn submit_state_and_proof(
     // prepare the input the contract call and the tx itself
     let proof: ParsedPlonkProof = proof.into();
     let new_state: ParsedLightClientState = public_input.into();
-    let tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    let mut tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    // TODO: (alex) centralize this to a dedicated ethtxmanager with gas control
+    // frugal gas price: set to SafeGasPrice based on live price oracle
+    // safe to be included with low priority position in a block.
+    // ignore this if etherscan fail to return
+    if let Ok(res) =
+        reqwest::get("https://api.etherscan.io/api?module=gastracker&action=gasoracle").await
+    {
+        // parse etherscan response,
+        // if etherscan API changes, we log the warning and fallback to default (by falling through)
+        if let Ok(res) = res.json::<EtherscanPriceOracleResponse>().await {
+            if let TypedTransaction::Eip1559(inner) = &mut tx.tx {
+                // both safe unwraps below assuming correct etherscan stream
+                let safe_max_fee: U256 = parse_units(res.result.safe_max_fee.clone(), "gwei")
+                    .unwrap()
+                    .into();
+                let base_fee: U256 = parse_units(res.result.base_fee.clone(), "gwei")
+                    .unwrap()
+                    .into();
+                inner.max_fee_per_gas = Some(safe_max_fee);
+                inner.max_priority_fee_per_gas = Some(safe_max_fee - base_fee);
```

Comment:
@alxiong It looks like this could underflow it would probably lead to the transaction being rejected because we don't have the funds but let's not rely on that.

Reply:
see if you like [6106914](https://github.com/EspressoSystems/espresso-sequencer/pull/2148/commits/6106914b3c282978b1e5a4036c548a51c0d117fa)

Reply:
this is outdated, see the latest commits that use blocknative and `trait PriceOracle` from ethers-rs